dunai/src/Data/MonadicStreamFunction/Instances/Profunctor.hs

```diff
@@ -0,0 +1,31 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+-- | Instances of 'Profunctor' and 'Traversing' for Monadic Stream Functions ('MSF').
+--
+--   Import this module to include the (orphan) instances.
+module Data.MonadicStreamFunction.Instances.Profunctor where
+
+-- base
+import Control.Arrow
+
+-- profunctors
+import Data.Profunctor
+import Data.Profunctor.Traversing
+
+-- dunai
+import Data.MonadicStreamFunction.Core
+import Data.MonadicStreamFunction.InternalCore
+import Data.MonadicStreamFunction.Instances.ArrowChoice
+
+instance Monad m => Choice (MSF m) where
+  left' = left
+
+instance Monad m => Strong (MSF m) where
+  first' = first
+
+instance Monad m => Profunctor (MSF m) where
+  dimap l r p = arr l >>> p >>> arr r
+
+instance (Monad m) => Traversing (MSF m) where
+  traverse' (MSF msf) = MSF $ \xs -> do
+    fbm <- mapM msf xs
```

Comment:
`fbm :: f (b, MSF m a b)`. So from that we need to return a tuple. The first half is easy and sensible. The second half confuses me. I would _like_ to use `fbm` to construct the 2nd half of the tuple, and `sequence` / `traverse` almost feel right, but, in the end I just reused the input parameter, which feels wrong, but type checks and seems to work. Thoughts?

Reply:
Yes, it can't be quite right because this way the internal state of the `MSF` will not advance. 

One strategy could be to recognise that the type of `unMSF :: MSF m a b -> a -> (b, MSF m a b)` is isomorphic to `unMSF' :: a -> StateT (MSF m a b) b`. Then the heartpiece is to do `traverse unMSF'`.

A similar approach is discussed here: https://github.com/turion/essence-of-live-coding/issues/57#issuecomment-876296306

Reply:
Let me know if you want more help with this :)

Reply:
Great intuition, that made it quite simple.

I added a custom `StateT` and then tried to back it back out (to just an unwrapped fn), because I don't want to add an mtl dependency.

But alas, this instance doesn't exist:

```haskell
instanceish Functor m => Applicative (s -> m (a, s)) where  -- Applicative in that `a`
```

Since we can newtype a `StateT` equivalent, I'd recommend doing that instead of bringing in mtl, unless there's a way that the `traverse ...` can get at the correct `Applicative` instance? I'll think on this some more.

Reply:
You don't need to add `mtl` as a dependency. `StateT` is defined in `transformers`, and `dunai` already has a dependency on it. See https://hackage.haskell.org/package/transformers-0.5.6.2/docs/Control-Monad-Trans-State-Strict.html. It has all the instances you need.

I'm not sure whether we need strict or lazy state. I'm tending to strict state.

Reply:
updated!

So I'm not sure I understand the tradeoff between strict and lazy state. I would guess that laziness might be necessary for ArrowLoop or MonadFix to work properly... But I'm not sure.